src/library_wasmfs.js

```diff
@@ -145,64 +141,54 @@ var WasmfsLibrary = {
     var len = lengthBytesUTF8(s) + 1;
     stringToUTF8(s, fileNameBuffer, len);
   },
-  _wasmfs_write_js_file: function(index, buffer, length, offset) {
-    try {
-      if (!wasmFS$JSMemoryFiles[index]) {
-        // Initialize typed array on first write operation.
-        wasmFS$JSMemoryFiles[index] = new Uint8Array(offset + length);
-      }
-
-      if (offset + length > wasmFS$JSMemoryFiles[index].length) {
-        // Resize the typed array if the length of the write buffer exceeds its capacity.
-        var oldContents = wasmFS$JSMemoryFiles[index];
-        var newContents = new Uint8Array(offset + length);
-        newContents.set(oldContents);
-        wasmFS$JSMemoryFiles[index] = newContents;
-      }
-
-      wasmFS$JSMemoryFiles[index].set(HEAPU8.subarray(buffer, buffer + length), offset);
-      return 0;
-    } catch (err) {
-      return {{{ cDefine('EIO') }}};
-    }
+  _wasmfs_get_preloaded_file_size: function(index) {
+    return wasmFS$preloadedFiles[index].fileData.length;
   },
-  _wasmfs_read_js_file: function(index, buffer, length, offset) {
-    try {
-      HEAPU8.set(wasmFS$JSMemoryFiles[index].subarray(offset, offset + length), buffer);
-      return 0;
-    } catch (err) {
-      return {{{ cDefine('EIO') }}};
-    }
+  _wasmfs_copy_preloaded_file_data: function(index, buffer) {
+    HEAPU8.set(wasmFS$preloadedFiles[index].fileData, buffer);
   },
-  _wasmfs_get_js_file_size: function(index) {
-    return wasmFS$JSMemoryFiles[index] ? wasmFS$JSMemoryFiles[index].length : 0;
+
+  // Backend support. wasmFS$backends will contain a mapping of backend IDs to
+  // the JS code that implements them. This is the JS side of the JSImpl class
+  // in C++, together with the js_impl calls defined right after it.
+  $wasmFS$backends: {},
+
+  _wasmfs_jsimpl_constructor: function(backend, file) {
+#if ASSERTIONS
+    assert(wasmFS$backends[backend]);
+#endif
+    return wasmFS$backends[backend].constructor(file);
```

Comment:
Sounds good. So we'll want something like `[new|delete][File|Directory|Symlink]`? And then a JS backend would implement just the ones it needs to just `newFile/deleteFile` for a basic JS File.

Thoughts on `new/delete`? This is about the lifecycle of the C++ File object, so I think new/delete might make sense. But could also be `alloc/free` or something else. `free` might be less likely to be confused with removing a file from the actual filesystem which `delete` might...

Reply:
I wonder if we want to expose the lifecycle of a conceptual file rather than the underlying C++ object. So instead of a destructor, maybe we should expose `onUnlink`, `onClose`, and `onDelete`. The C++ destructor would call `onDelete` only if the file is actually being deleted and not when the program is shutting down.

But a problem with that is that backends may actually want to know when the underlying C++ object is being destructed, perhaps because they own other C++ objects that should also be destructed at that point. So maybe we need a destructor hook in addition to the file lifecycle hooks? In that case, it is probably sufficient for this PR to just have the destructor hook.

In that case, I would want to avoid using `delete` to describe the destructor hook, since it would be ambiguous whether `delete` referred to the in-memory resources or the file itself, as you mentioned.

`newFile/freeFile` or `createFile/freeFile` or `constructFile/destructFile` might be good names. I don't feel strongly, except that I would avoid `delete`.

Reply:
SGTM 👍